File: src/main/scala/jk_5/nailed/server/mappack/NailedMappackRegistry.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.mappack

import jk_5.nailed.api.Platform
import jk_5.nailed.api.event.mappack.{MappackRegisteredEvent, MappackUnregisteredEvent}
import jk_5.nailed.api.mappack.{Mappack, MappackRegistry}
import jk_5.nailed.server.NailedPlatform

import scala.collection.mutable

/**
 * No description given
 *
 * @author jk-5
 */
object NailedMappackRegistry extends MappackRegistry {

  private val mappacks = mutable.HashMap[String, Mappack]()

  override def register(mappack: Mappack): Boolean =
    if(this.mappacks.exists(p => p._1 == mappack.getId || p._2 == mappack)) false
    else{
      this.mappacks.put(mappack.getId, mappack)
      NailedPlatform.globalEventBus.post(new MappackRegisteredEvent(mappack))
      true
    }

  override def getByName(name: String): Mappack = this.mappacks.get(name).orNull

  override def getByType[T](cl: Class[_ <: T]): java.util.Collection[T] = {
    java.util.Arrays.asList(this.mappacks.values.collect {
      case special if cl.isAssignableFrom(special.getClass) => special
      case _ =>
    }.toArray.asInstanceOf[Array[T]]: _*)
  }

  override def getAll: java.util.Collection[Mappack] = java.util.Arrays.asList(this.mappacks.values.toArray: _*)
  override def getAllIds: java.util.Collection[String] = java.util.Arrays.asList(this.mappacks.values.toArray.map(_.getId): _*)

  override def unregister(mappack: Mappack): Boolean = {
    if(!this.mappacks.exists(p => p._1 == mappack.getId && p._2 == mappack)) false
    else{
      this.mappacks.remove(mappack.getId)
      NailedPlatform.globalEventBus.post(new MappackUnregisteredEvent(mappack))
      true
    }
  }
}

trait MappackRegistryTrait extends Platform {
  override def getMappackRegistry = NailedMappackRegistry
}

File: src/main/java/jk_5/nailed/server/mappack/NailedMappackRegistry.java
package jk_5.nailed.server.mappack;

import com.google.common.collect.ImmutableList;
import jk_5.nailed.api.event.mappack.MappackRegisteredEvent;
import jk_5.nailed.api.event.mappack.MappackUnregisteredEvent;
import jk_5.nailed.api.event.mappack.RegisterMappacksEvent;
import jk_5.nailed.api.mappack.Mappack;
import jk_5.nailed.api.mappack.MappackRegistry;
import jk_5.nailed.server.NailedEventFactory;
import jk_5.nailed.server.map.NailedMapLoader;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.util.*;

public class NailedMappackRegistry implements MappackRegistry {

    private static final NailedMappackRegistry INSTANCE = new NailedMappackRegistry();
    private static final Logger logger = LogManager.getLogger();

    private final Map<String, Mappack> mappacks = new HashMap<String, Mappack>();

    @Override
    public boolean register(@Nonnull Mappack mappack) {
        if(mappacks.containsKey(mappack.getId()) || mappacks.containsValue(mappack)){
            return false;
        }
        this.mappacks.put(mappack.getId(), mappack);
        NailedEventFactory.fireEvent(new MappackRegisteredEvent(mappack));
        return true;
    }

    @Nullable
    @Override
    public Mappack getByName(@Nonnull String name) {
        return this.mappacks.get(name);
    }

    @Nonnull
    @Override
    public <T> Collection<T> getByType(@Nonnull Class<? extends T> cl) {
        List<T> ret = new ArrayList<T>();
        for (Mappack mappack : this.mappacks.values()) {
            if(cl.isAssignableFrom(mappack.getClass())){
                ret.add((T) mappack);
            }
        }
        return ret;
    }

    @Nonnull
    @Override
    public Collection<Mappack> getAll() {
        return ImmutableList.copyOf(this.mappacks.values());
    }

    @Nonnull
    @Override
    public Collection<String> getAllIds() {
        return ImmutableList.copyOf(this.mappacks.keySet());
    }

    @Override
    public boolean unregister(@Nonnull Mappack mappack) {
        if(!mappacks.containsKey(mappack.getId()) && !mappacks.containsValue(mappack)){
            return false;
        }
        this.mappacks.remove(mappack.getId());
        NailedEventFactory.fireEvent(new MappackUnregisteredEvent(mappack));
        return true;
    }

    public void reload(){
        logger.info("Reloading mappacks");
        mappacks.clear();
        NailedEventFactory.fireEvent(new RegisterMappacksEvent(this, NailedMapLoader.instance()));
        NailedMapLoader.instance().checkLobbyMappack();
    }

    public static NailedMappackRegistry instance(){
        return INSTANCE;
    }
}
